File: server/app/models/artifact.py
"""Artifact model."""

import uuid
from datetime import datetime
from enum import Enum as PyEnum

from sqlalchemy import BigInteger, Column, DateTime, ForeignKey, String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.db.base import Base
from app.db.types import DialectJSON


class ArtifactType(PyEnum):
    """Artifact type enumeration."""

    PDF = "pdf"
    IR_V1 = "ir_v1"
    IR_V2 = "ir_v2"
    MUSICXML = "musicxml"
    MIDI = "midi"
    SVG = "svg"


class Artifact(Base):
    """Artifact model for storing file metadata."""

    __tablename__ = "artifacts"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)
    job_id = Column(UUID(as_uuid=True), ForeignKey("jobs.id"), nullable=False, index=True)
    artifact_type = Column(String(50), nullable=False, index=True)
    schema_version = Column(String(20), nullable=False, default="1.0.0")
    storage_path = Column(String(512), nullable=False)
    file_size = Column(BigInteger, nullable=False)
    checksum = Column(String(64), nullable=False)  # SHA256 hex
    artifact_metadata = Column(DialectJSON, default=dict, nullable=False)
    parent_artifact_id = Column(UUID(as_uuid=True), ForeignKey("artifacts.id"), nullable=True)
    created_at = Column(DateTime(timezone=True), default=datetime.utcnow, nullable=False)

    # Relationships
    job = relationship("Job", back_populates="artifacts")
    parent_artifact = relationship("Artifact", remote_side=[id], backref="child_artifacts")

    def __repr__(self) -> str:
        return f"<Artifact(id={self.id}, type={self.artifact_type}, job_id={self.job_id})>"

File: client/lib/data/models/artifact.dart
import 'package:json_annotation/json_annotation.dart';

part 'artifact.g.dart';

enum ArtifactType {
  @JsonValue('pdf')
  pdf,
  @JsonValue('ir_v1')
  irV1,
  @JsonValue('ir_v2')
  irV2,
  @JsonValue('musicxml')
  musicxml,
  @JsonValue('midi')
  midi,
  @JsonValue('svg')
  svg,
  @JsonValue('png')
  png,
}

@JsonSerializable()
class Artifact {
  final String id;
  @JsonKey(name: 'job_id')
  final String jobId;
  @JsonKey(name: 'artifact_type')
  final String artifactType;
  @JsonKey(name: 'schema_version')
  final String? schemaVersion;
  @JsonKey(name: 'storage_path')
  final String storagePath;
  @JsonKey(name: 'file_size')
  final int fileSize;
  final String checksum;
  @JsonKey(name: 'parent_artifact_id')
  final String? parentArtifactId;
  @JsonKey(name: 'artifact_metadata')
  final Map<String, dynamic>? artifactMetadata;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  
  Artifact({
    required this.id,
    required this.jobId,
    required this.artifactType,
    this.schemaVersion,
    required this.storagePath,
    required this.fileSize,
    required this.checksum,
    this.parentArtifactId,
    this.artifactMetadata,
    required this.createdAt,
  });
  
  factory Artifact.fromJson(Map<String, dynamic> json) => 
      _$ArtifactFromJson(json);
  Map<String, dynamic> toJson() => _$ArtifactToJson(this);
  
  ArtifactType get typeEnum {
    // Map backend snake_case to Dart enum names
    final typeMapping = {
      'pdf': ArtifactType.pdf,
      'ir_v1': ArtifactType.irV1,
      'ir_v2': ArtifactType.irV2,
      'musicxml': ArtifactType.musicxml,
      'midi': ArtifactType.midi,
      'svg': ArtifactType.svg,
      'png': ArtifactType.png,
    };
    return typeMapping[artifactType.toLowerCase()] ?? ArtifactType.pdf;
  }
}

